File: M08_Architecture/C02_Component/S01_ViewModel/src/main/java/net/bi4vmr/study/lifecycle/MyViewModel2.java
package net.bi4vmr.study.lifecycle;

import android.util.Log;

import androidx.lifecycle.ViewModel;

import java.util.UUID;

/**
 * 自定义ViewModel：生命周期测试。
 *
 * @author bi4vmr@outlook.com
 * @since 1.0.0
 */
public class MyViewModel2 extends ViewModel {

    private static final String TAG = "TestApp-" + MyViewModel2.class.getSimpleName();

    // 实例ID
    public final String id;

    public MyViewModel2() {
        // 生成随机ID，标识当前实例。
        id = genRandomID();
        Log.i(TAG, "VM created. ID:[" + id + "]");
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        Log.i(TAG, "OnCleared. ID:[" + id + "]");
    }

    // 获取随机ID
    private String genRandomID() {
        return UUID.randomUUID()
                .toString()
                .toUpperCase()
                .substring(0, 6);
    }
}

File: M03_UI/C10_Compose/S04_State/src/main/kotlin/net/bi4vmr/study/statesave/MyViewModel2.kt
package net.bi4vmr.study.statesave

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel

/**
 * ViewModel：计数器示例。
 *
 * @author bi4vmr@outlook.com
 * @since 1.0.0
 */
class MyViewModel2 : ViewModel() {

    // 可变 LiveData ，仅内部使用。
    private val _count: MutableLiveData<Int> = MutableLiveData(0)

    // 不可变 LiveData ，与可变实例保持同步，供外部观察。
    val count: LiveData<Int> = _count


    // 业务接口：数值增加。
    fun add() {
        _count.value = _count.value?.plus(1)
    }

    // 业务接口：数值减少。
    fun reduce() {
        _count.value = _count.value?.minus(1)
    }
}
